src/entry.rs

```diff
@@ -42,9 +105,66 @@ unsafe extern "C" fn entry() -> ! {
         // Prepare the stack.
         "adr_l x30, boot_stack_end",
         "mov sp, x30",
+        // Perform final Rust entrypoint setup
+        "b {entry_prepare_image}",
+        ".purgem adr_l",
+        entry_prepare_image = sym entry_prepare_image
+    )
+}
+
+#[cfg(not(feature = "relocate"))]
+#[unsafe(naked)]
+unsafe extern "C" fn entry_prepare_image() -> ! {
+    naked_asm!(
+        // Call into Rust code.
+        "b {rust_entry}",
+        rust_entry = sym rust_entry
+    )
+}
+
+#[cfg(feature = "relocate")]
+#[unsafe(naked)]
+unsafe extern "C" fn entry_prepare_image() -> ! {
+    naked_asm!(
+        ".macro adr_l, reg:req, sym:req",
+        r"adrp \reg, \sym",
+        r"add \reg, \reg, :lo12:\sym",
+        ".endm",
+        // Where the image was loaded
+        "adr_l x7, text_begin",
+        // Start relocating.
+        // let mut rela: *mut Elf64Rela = __rela_start;
+        "adr_l x9, __rela_start",
+        // let rela_end: *mut Elf64Rela = __rela_end;
+        "adr_l x8, __rela_end",
+        "b 1f",
+        "2:",
+        // rela = rela.wrapping_offset(1);
+        "add x9, x9, #24",
+        "1:",
+        // while rela < rela_end {
+        "cmp x9, x8",
+        "b.hs 3f",
+        // let r_type = unsafe { *rela }.r_info & 0xffff_ffff;
+        "ldr w10, [x9, #8]",
+        // if r_type != R_AARCH64_RELATIVE { continue };
+        "cmp w10, #1027",
+        "b.ne 2b",
+        // let mut new_ptr = unsafe { *rela }.r_addend;
+        "ldr x10, [x9, #16]",
+        // let r_offset = unsafe { *rela }.r_offset;
+        "ldr x11, [x9]",
+        //  new_ptr += offset;
```

Comment:
You can simplify this a bit, and remove the need for the jump at the start of the loop, by doing something like

```asm
ldp x10, x11, [x9], #24 
ldr x12, [x9, #-8]
```

Note that `R_AARCH64_RELATIVE` is the only RELA type that can reasonably expected to occur here, so performing the load of the offset and addend fields unconditionally will not make any difference.